File: code/src/main/kotlin/cn/dong/leetcode/TwoSum.java
package cn.dong.leetcode;

import java.util.Arrays;
import java.util.HashMap;

/**
 * https://leetcode.com/problems/two-sum/
 *
 * @author dong on 2018/10/30.
 */
public class TwoSum {
    public static int[] twoSum(int[] nums, int target) {
        if (nums == null) return new int[]{};
        HashMap<Integer, Integer> map = new HashMap<>(); // <value, index>
        for (int i = 0; i < nums.length; i++) {
            Integer index = map.get(target - nums[i]);
            if (index != null) {
                return new int[]{i, index};
            }
            map.put(nums[i], i);
        }
        return new int[]{};
    }

    public static void main(String[] args) {
        int[] input = new int[]{2, 7, 11, 15};
        int[] output = twoSum(input, 9);
        System.out.println(Arrays.toString(output));
    }
}

File: code/src/main/kotlin/cn/dong/leetcode/1/two-sum.kt
package cn.dong.leetcode.`1`

/**
 * [1. 两数之和 - 力扣（LeetCode）](https://leetcode.cn/problems/two-sum)
 */
class Solution_1 {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val map = HashMap<Int, Int>() // num, Index

        for (i in 0..nums.size -1 ) {

            val num = nums[i]

            val expectedNum = target - num
            val index = map[expectedNum]
            if (index != null) {
                return intArrayOf(index, i)
            } else {
                map[num] = i
            }
        }
        return intArrayOf(0)
    }
}